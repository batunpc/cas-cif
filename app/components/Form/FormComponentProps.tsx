import type { AjvError } from "@rjsf/core";
export default interface FormComponentProps {
  formData: any;
  onChange: (changeObject: object) => void;
  onFormErrors: (errorsArray: AjvError[]) => void;
}
