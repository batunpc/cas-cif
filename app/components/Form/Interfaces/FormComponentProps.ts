import { FormProps } from "@rjsf/core";

type FormComponentProps<T = any> = Partial<FormProps<T>>;

export default FormComponentProps;
