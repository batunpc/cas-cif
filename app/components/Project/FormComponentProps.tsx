export default interface FormComponentProps {
  formData: any;
  applyChangesFromComponent: (changeObject: object) => void;
  onFormErrors: (errorsObject: object) => void;
}
