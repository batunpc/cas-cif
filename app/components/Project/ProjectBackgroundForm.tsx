import type { JSONSchema7 } from "json-schema";
import React from "react";
import FormBase from "./FormBase";
import FormComponentProps from "./FormComponentProps";

const schema: JSONSchema7 = {
  type: "object",
  required: ["rfp_number", "description"],
  properties: {
    rfp_number: {
      type: "string",
      title: "RFP Number",
      pattern: "^((\\d{4})-RFP-([1-2])-(\\d{3,4})-([A-Z]{4}))$",
    },
    description: { type: "string", title: "Description" },
  },
};

const uiSchema = {
  rfp_number: {
    "ui:placeholder": "2020-RFP-1-456-ABCD",
    "ui:col-md": 12,
    "bcgov:size": "small",
  },
  description: {
    "ui:placeholder": "describe the project...",
    "ui:col-md": 12,
    "bcgov:size": "small",
  },
};

const createInitialData = (formData) => {
  return {
    rfp_number: formData.rfp_number || undefined,
    description: formData.description || undefined,
  };
};

const ProjectBackgroundForm: React.FunctionComponent<FormComponentProps> = (
  props
) => {
  return (
    <FormBase
      {...props}
      createInitialData={createInitialData}
      schema={schema}
      uiSchema={uiSchema}
    />
  );
};

export default ProjectBackgroundForm;
