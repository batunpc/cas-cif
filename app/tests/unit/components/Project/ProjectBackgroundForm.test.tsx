import { render, screen, fireEvent } from "@testing-library/react";
import FormComponentProps from "components/Project/FormComponentProps";
import ProjectBackgroundForm from "components/Project/ProjectBackgroundForm";

describe("The Project Background Form", () => {
  it("matches the snapshot", () => {
    const props: FormComponentProps = {
      formData: {},
      applyChangesFromComponent: jest.fn(),
      onFormErrors: jest.fn(),
    };

    const componentUnderTest = render(<ProjectBackgroundForm {...props} />);

    expect(componentUnderTest.container).toMatchSnapshot();
  });
  it("triggers the applyFormChange with the proper data", () => {
    const applySpy = jest.fn();

    const props: FormComponentProps = {
      formData: {},
      applyChangesFromComponent: applySpy,
      onFormErrors: jest.fn(),
    };

    render(<ProjectBackgroundForm {...props} />);

    fireEvent.change(screen.getByLabelText("RFP Number*"), {
      target: { value: "testidentifier" },
    });

    expect(applySpy).toHaveBeenCalledWith({
      rfp_number: "testidentifier",
      description: undefined,
    });
    applySpy.mockClear();

    fireEvent.change(screen.getByLabelText("Description*"), {
      target: { value: "testdescription" },
    });

    expect(applySpy).toHaveBeenCalledWith({
      rfp_number: "testidentifier",
      description: "testdescription",
    });
  });
  it("loads with the correct initial form data", () => {
    const props: FormComponentProps = {
      formData: {
        rfp_number: "12345678",
        description: "d",
      },
      applyChangesFromComponent: jest.fn(),
      onFormErrors: jest.fn(),
    };

    render(<ProjectBackgroundForm {...props} />);

    expect(screen.getByLabelText("RFP Number*").value).toBe("12345678");
    expect(screen.getByLabelText("Description*").value).toBe("d");
  });
  it("calls onformerrors on first render if there are errors", () => {
    const onFormErrorsSpy = jest.fn();

    const props: FormComponentProps = {
      formData: {
        rfp_number: "12345678",
        description: "d",
      },
      applyChangesFromComponent: jest.fn(),
      onFormErrors: onFormErrorsSpy,
    };

    render(<ProjectBackgroundForm {...props} />);

    expect(onFormErrorsSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        rfp_number: expect.anything(),
        description: null,
      })
    );
  });

  it("calls onformerrors with null if there are no errors", () => {
    const onFormErrorsSpy = jest.fn();

    const props: FormComponentProps = {
      formData: {
        rfp_number: "1999-RFP-1-123-ABCD",
        description: "d",
      },
      applyChangesFromComponent: jest.fn(),
      onFormErrors: onFormErrorsSpy,
    };

    render(<ProjectBackgroundForm {...props} />);

    expect(onFormErrorsSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        rfp_number: null,
        description: null,
      })
    );
  });

  it("calls onformerrors if a fields becomes empty", () => {
    const onFormErrorsSpy = jest.fn();

    const props: FormComponentProps = {
      formData: {
        rfp_number: "1999-RFP-1-123-ABCD",
        description: "desc",
      },
      applyChangesFromComponent: jest.fn(),
      onFormErrors: onFormErrorsSpy,
    };

    render(<ProjectBackgroundForm {...props} />);

    fireEvent.change(screen.getByLabelText("Description*"), {
      target: { value: "" },
    });

    expect(onFormErrorsSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        rfp_number: null,
        description: { __errors: ["is a required property"] },
      })
    );
  });

  it("calls onformerrors if the project unique id doesnt match format", () => {
    const onFormErrorsSpy = jest.fn();

    const props: FormComponentProps = {
      formData: {
        rfp_number: "1999123-RFP-1-123-ABCD",
        description: "desc",
      },
      applyChangesFromComponent: jest.fn(),
      onFormErrors: onFormErrorsSpy,
    };

    render(<ProjectBackgroundForm {...props} />);

    expect(onFormErrorsSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        rfp_number: expect.anything(),
        description: null,
      })
    );
  });
});
