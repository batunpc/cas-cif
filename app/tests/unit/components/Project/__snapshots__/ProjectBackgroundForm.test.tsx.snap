// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Project Background Form matches the snapshot 1`] = `
<div>
  <fieldset
    class="jsx-3408663875"
  >
    <legend
      class="jsx-3408663875"
    >
      Background
    </legend>
    <form
      class="rjsf"
    >
      <div
        class="panel panel-danger errors"
      >
        <div
          class="panel-heading"
        >
          <h3
            class="panel-title"
          >
            Errors
          </h3>
        </div>
        <ul
          class="list-group"
        >
          <li
            class="list-group-item text-danger"
          >
            .rfp_number is a required property
          </li>
          <li
            class="list-group-item text-danger"
          >
            .description is a required property
          </li>
        </ul>
      </div>
      <fieldset
        class="sc-dlfnuX hBsleh pg-fieldset"
        id="root"
      >
        <div
          class="sc-hBEYId iNUeqH pg-input"
        >
          <label
            class="sc-fodVek ensGkx pg-input-label"
            for="root_rfp_number"
          >
            RFP Number*
          </label>
          <input
            aria-label="RFP Number*"
            class="sc-fFucqa ljMVny pg-input-input"
            id="root_rfp_number"
            name="root_rfp_number-input"
            placeholder="2020-RFP-1-456-ABCD"
            required=""
            value=""
          />
        </div>
        <div>
          <ul
            class="error-detail bs-callout bs-callout-info"
          >
            <li
              class="text-danger"
            >
              is a required property
            </li>
          </ul>
        </div>
        <div
          class="sc-hBEYId iNUeqH pg-input"
        >
          <label
            class="sc-fodVek ensGkx pg-input-label"
            for="root_description"
          >
            Description*
          </label>
          <input
            aria-label="Description*"
            class="sc-fFucqa ljMVny pg-input-input"
            id="root_description"
            name="root_description-input"
            placeholder="describe the project..."
            required=""
            value=""
          />
        </div>
        <div>
          <ul
            class="error-detail bs-callout bs-callout-info"
          >
            <li
              class="text-danger"
            >
              is a required property
            </li>
          </ul>
        </div>
      </fieldset>
    </form>
  </fieldset>
</div>
`;
